### ISSUE TITLE
apf::receiveAllCopies(apf::Mesh *): Assertion `a == b' failed.

### ISSUE THREAD

**Issue by Thomas-Ulrich** [SCOREC/core#23]
Hi, I using PUMGEN, a code written by Sebastian Rettenberger, which uses SCOREC combine with Simulation Modeling Suite (Release 10.0-160429), to produce 3d tetrahedral meshes, partition them and write them in a custom netcdf format.
(here is the source https://github.com/TUM-I5/PUML)
I m running into crashes that I think are related to SCOREC. Any idea why?

Thanks in advance,

Thomas Ulrich,
PhD student, Geophysics, LMU Munich.

here is how I run the program:

pumgen -s simmodsuite -l SimModelerLib.lic --mesh "Mesh case 1" --analysis "Analysis case 1" --sim_log logtpv16.dat tpv16.smd 28 trash.28.nc

here is the program output:

Wed May 11 11:53:53, Info:  Using SimModSuite
Wed May 11 11:53:57, Info:  Loading model
Wed May 11 11:53:58, Info:  Extracting cases
Wed May 11 11:53:58, Info:  Starting the surface mesher
Wed May 11 11:54:02, Info:  Starting the volume mesher
Wed May 11 11:55:39, Info:  Converting mesh to APF

here is the error log:

pumgen: /gpfs/work/h019z/di73yeq/core-sim/apf/apfVerify.cc:256: void apf::receiveAllCopies(apf::Mesh *): Assertion `a == b' failed.
Abort(1) on node 18 (rank 18 in comm 1140850688): application called MPI_Abort(MPI_COMM_WORLD, 1) - process 18
Thread 11 (Thread 0x2b36fc7f6700 (LWP 35889)):
#0  0x00002b36f6ea9ce3 in epoll_wait () from /lib64/libc.so.6
#1  0x00002b36fc194436 in poe_exiting_thread () from /opt/ibmhpc/pe1402/base/intel/lib64/libpoe.so
#2  0x00002b36f5ec9806 in start_thread () from /lib64/libpthread.so.0
#3  0x00002b36f6ea965d in clone () from /lib64/libc.so.6
#4  0x0000000000000000 in ?? ()
Thread 10 (Thread 0x2b36fc9f7700 (LWP 35890)):
#0  0x00002b36f6ea9ce3 in epoll_wait () from /lib64/libc.so.6
#1  0x00002b36fc192bb3 in pm_child_sig_thread () from /opt/ibmhpc/pe1402/base/intel/lib64/libpoe.so
#2  0x00002b36f5ec9806 in start_thread () from /lib64/libpthread.so.0
#3  0x00002b36f6ea965d in clone () from /lib64/libc.so.6
#4  0x0000000000000000 in ?? ()
Thread 9 (Thread 0x2b36fcbf8700 (LWP 35891)):
#0  0x00002b36f5ed1527 in do_sigwait () from /lib64/libpthread.so.0
#1  0x00002b36f5ed15cd in sigwait () from /lib64/libpthread.so.0
#2  0x00002b36fc193ba2 in pm_async_thread () from /opt/ibmhpc/pe1402/base/intel/lib64/libpoe.so
#3  0x00002b36f5ec9806 in start_thread () from /lib64/libpthread.so.0
#4  0x00002b36f6ea965d in clone () from /lib64/libc.so.6
#5  0x0000000000000000 in ?? ()
Thread 8 (Thread 0x2b36ff042700 (LWP 35897)):
#0  0x00002b36f5ecd66c in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
#1  0x00002b36fda8e58b in hal_ibl_user_intr_hndlr () from /opt/ibmhpc/pe1402/base/intel/lib64/libhal64_ibm.so
#2  0x00002b36f5ec9806 in start_thread () from /lib64/libpthread.so.0
#3  0x00002b36f6ea965d in clone () from /lib64/libc.so.6
#4  0x0000000000000000 in ?? ()
Thread 7 (Thread 0x2b36ff243700 (LWP 35898)):
#0  0x00002b36f6ea9ce3 in epoll_wait () from /lib64/libc.so.6
#1  0x00002b36fda8e183 in hal_ibl_async_intr_hndlr () from /opt/ibmhpc/pe1402/base/intel/lib64/libhal64_ibm.so
#2  0x00002b36f5ec9806 in start_thread () from /lib64/libpthread.so.0
#3  0x00002b36f6ea965d in clone () from /lib64/libc.so.6
#4  0x0000000000000000 in ?? ()
Thread 6 (Thread 0x2b370bb49700 (LWP 35996)):
#0  0x00002b36f5ecd66c in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
#1  0x00002b36f9b3a3a3 in shm_dispatcher_thread (arg=0x2b36ffa561ac) at /project/sprelbrew/build/rbrews002a/src/ppe/lapi/lapi_shm.c:2282
#2  0x00002b36f5ec9806 in start_thread () from /lib64/libpthread.so.0
#3  0x00002b36f6ea965d in clone () from /lib64/libc.so.6
#4  0x0000000000000000 in ?? ()
Thread 5 (Thread 0x2b370bff6700 (LWP 35997)):
#0  0x00002b36f5ecd66c in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
#1  0x00002b36f9b5236f in rc_ibl_intr_hndlr (param=0x2b36fa16315c <intr_hndlr_info+60>) at /project/sprelbrew/build/rbrews002a/src/ppe/lapi/lapi_rc_rdma_verbs_wrappers.c:1180
#2  0x00002b36f5ec9806 in start_thread () from /lib64/libpthread.so.0
#3  0x00002b36f6ea965d in clone () from /lib64/libc.so.6
#4  0x0000000000000000 in ?? ()
Thread 4 (Thread 0x2b370c1f7700 (LWP 35998)):
#0  0x00002b36f6ea9ce3 in epoll_wait () from /lib64/libc.so.6
#1  0x00002b36f9b527b1 in rc_ibl_async_intr_hndlr (param=0x16) at /project/sprelbrew/build/rbrews002a/src/ppe/lapi/lapi_rc_rdma_verbs_wrappers.c:1374
#2  0x00002b36f5ec9806 in start_thread () from /lib64/libpthread.so.0
#3  0x00002b36f6ea965d in clone () from /lib64/libc.so.6
#4  0x0000000000000000 in ?? ()
Thread 3 (Thread 0x2b370c3f8700 (LWP 35999)):
#0  0x00002b36f5ecd9fc in pthread_cond_timedwait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
#1  0x00002b36f9b1c37f in _timer_arm (timer=0x1800c44) at /project/sprelbrew/build/rbrews002a/src/ppe/lapi/intrhndlrs.c:340
#2  0x00002b36f9b1c0e4 in _lapi_tmr_thrd (param=0x1800c44) at /project/sprelbrew/build/rbrews002a/src/ppe/lapi/intrhndlrs.c:541
#3  0x00002b36f5ec9806 in start_thread () from /lib64/libpthread.so.0
#4  0x00002b36f6ea965d in clone () from /lib64/libc.so.6
#5  0x0000000000000000 in ?? ()
Thread 2 (Thread 0x2b370c5f9700 (LWP 36113)):
#0  0x00002b36f6ea0186 in poll () from /lib64/libc.so.6
#1  0x00002b36fbf5e642 in Connection::Wait (this=0x18117f0) at /project/sprelbrew/build/rbrews002a/src/ppe/pnsd/connection.cpp:169
#2  0x00002b36fbf4a3ce in internal_pnsd_api_wait_for_updates (handle=<optimized out>, wakeup_event=0x2b370c5ea874, device_name=<optimized out>, adapter_type=<optimized out>, win_id=<optimized out>, cmd_string=0x2b370c5ea860, opt_length=0x2b370c5ea870, opt=0x2b370c5ea868) at /project/sprelbrew/build/rbrews002a/src/ppe/pnsd/pnsd_api.cpp:337
#3  0x00002b36fbf4a6ac in pnsd_api_wait_for_updates (handle=207529648, wakeup_event_OUT=0x2, cmd_string_OUT=<optimized out>, opt_length=<optimized out>, opt_OUT=<optimized out>) at /project/sprelbrew/build/rbrews002a/src/ppe/pnsd/pnsd_api.cpp:379
#4  0x00002b36f9b410ab in preempt_monitor_thread (param=0x2b370c5ea6b0) at /project/sprelbrew/build/rbrews002a/src/ppe/lapi/lapi_preempt.c:523
#5  0x00002b36f5ec9806 in start_thread () from /lib64/libpthread.so.0
#6  0x00002b36f6ea965d in clone () from /lib64/libc.so.6
#7  0x0000000000000000 in ?? ()
Thread 1 (Thread 0x2b36fbf381e0 (LWP 35843)):
#0  0x00002b36f6e75b8f in waitpid () from /lib64/libc.so.6
#1  0x00002b36f6e09cc1 in do_system () from /lib64/libc.so.6
#2  0x00002b36f6e0a06c in system () from /lib64/libc.so.6
#3  0x00002b36fc19473f in pm_linux_print_coredump () from /opt/ibmhpc/pe1402/base/intel/lib64/libpoe.so
#4  0x00002b36fc193fda in pm_lwcf_signal_handler () from /opt/ibmhpc/pe1402/base/intel/lib64/libpoe.so
#5  <signal handler called>
#6  0x00002b36f6dfd885 in raise () from /lib64/libc.so.6
#7  0x00002b36f6dfee61 in abort () from /lib64/libc.so.6
#8  0x00002b36f6df6740 in __assert_fail () from /lib64/libc.so.6
#9  0x0000000000543cdc in apf::receiveAllCopies (m=0x8c03) at /gpfs/work/h019z/di73yeq/core-sim/apf/apfVerify.cc:256
#10 0x0000000000544991 in apf::verify (m=0x8c03) at /gpfs/work/h019z/di73yeq/core-sim/apf/apfVerify.cc:273
#11 0x00000000004540e7 in main (argc=14, argv=0x7ffde79d4fd8) at src/tools/pumgen.cpp:223
Abort(1) on node 0 (rank 0 in comm 1140850688): application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0
Abort(1) on node 1 (rank 1 in comm 1140850688): application called MPI_Abort(MPI_COMM_WORLD, 1) - process 1
Abort(1) on node 2 (rank 2 in comm 1140850688): application called MPI_Abort(MPI_COMM_WORLD, 1) - process 2
Abort(1) on node 3 (rank 3 in comm 1140850688): application called MPI_Abort(MPI_COMM_WORLD, 1) - process 3
Abort(1) on node 4 (rank 4 in comm 1140850688): application called MPI_Abort(MPI_COMM_WORLD, 1) - process 4
Abort(1) on node 5 (rank 5 in comm 1140850688): application called MPI_Abort(MPI_COMM_WORLD, 1) - process 5
Abort(1) on node 6 (rank 6 in comm 1140850688): application called MPI_Abort(MPI_COMM_WORLD, 1) - process 6
Abort(1) on node 7 (rank 7 in comm 1140850688): application called MPI_Abort(MPI_COMM_WORLD, 1) - process 7
Abort(1) on node 8 (rank 8 in comm 1140850688): application called MPI_Abort(MPI_COMM_WORLD, 1) - process 8
Abort(1) on node 9 (rank 9 in comm 1140850688): application called MPI_Abort(MPI_COMM_WORLD, 1) - process 9
Abort(1) on node 10 (rank 10 in comm 1140850688): application called MPI_Abort(MPI_COMM_WORLD, 1) - process 10
Abort(1) on node 11 (rank 11 in comm 1140850688): application called MPI_Abort(MPI_COMM_WORLD, 1) - process 11
Abort(1) on node 12 (rank 12 in comm 1140850688): application called MPI_Abort(MPI_COMM_WORLD, 1) - process 12
Abort(1) on node 13 (rank 13 in comm 1140850688): application called MPI_Abort(MPI_COMM_WORLD, 1) - process 13
Abort(1) on node 14 (rank 14 in comm 1140850688): application called MPI_Abort(MPI_COMM_WORLD, 1) - process 14
Abort(1) on node 15 (rank 15 in comm 1140850688): application called MPI_Abort(MPI_COMM_WORLD, 1) - process 15
Abort(1) on node 16 (rank 16 in comm 1140850688): application called MPI_Abort(MPI_COMM_WORLD, 1) - process 16
Abort(1) on node 17 (rank 17 in comm 1140850688): application called MPI_Abort(MPI_COMM_WORLD, 1) - process 17
Abort(1) on node 20 (rank 20 in comm 1140850688): application called MPI_Abort(MPI_COMM_WORLD, 1) - process 20
Abort(1) on node 21 (rank 21 in comm 1140850688): application called MPI_Abort(MPI_COMM_WORLD, 1) - process 21
Abort(1) on node 22 (rank 22 in comm 1140850688): application called MPI_Abort(MPI_COMM_WORLD, 1) - process 22
Abort(1) on node 23 (rank 23 in comm 1140850688): application called MPI_Abort(MPI_COMM_WORLD, 1) - process 23
Abort(1) on node 24 (rank 24 in comm 1140850688): application called MPI_Abort(MPI_COMM_WORLD, 1) - process 24
Abort(1) on node 25 (rank 25 in comm 1140850688): application called MPI_Abort(MPI_COMM_WORLD, 1) - process 25
Abort(1) on node 26 (rank 26 in comm 1140850688): application called MPI_Abort(MPI_COMM_WORLD, 1) - process 26
Abort(1) on node 27 (rank 27 in comm 1140850688): application called MPI_Abort(MPI_COMM_WORLD, 1) - process 27
Abort(1) on node 19 (rank 19 in comm 1140850688): application called MPI_Abort(MPI_COMM_WORLD, 1) - process 19

and here is the log from SimModeler library:

-- MS_init()
--[[
10.0-160429 Platform x64_rhel5_gcc41
--]]
MS_init()
-- SimDiscrete_start(0)
--[[
SimDiscrete: 160429 Platform: x64_rhel5_gcc41
--]]
SimDiscrete_start(0)
-- SimParasolid_start(1)
--[[
SimParasolid: 160429 Platform: x64_rhel5_gcc41 Version: 27_0_142
--]]
SimParasolid_start(1)
-- GM_load("tpv16.smd", nil, nil)
r28804384 = GM_load("tpv16.smd", nil, nil)
-- GM_nativeModel(r28804384)
nil = GM_nativeModel(r28804384)
--[[
pl27460696 = PList_new()
GM_isValid(r28804384, 0, pl27460696)
PList_delete(pl27460696)
--]]
pl27460696 = PList_new()
GM_isValid(r28804384, 0, pl27460696)
PList_delete(pl27460696)
-- MS_newMeshCase(r28804384)
n27803872 = MS_newMeshCase(r28804384)
-- AttCase_setModel(AttNode_byTag(r29094432, 8), r28804384)
AttCase_setModel(AttNode_byTag(r29094432, 8), r28804384)
-- AttCase_loadedModel(AttNode_byTag(r29094432, 8))
r28804384 = AttCase_loadedModel(AttNode_byTag(r29094432, 8))
-- MS_addCubeRefinement(n27803872, 1000, {0, 0, -10000}, {20000, 0, 0}, {0, 10000, 0}, {0, 0, 10000})
MS_addCubeRefinement(n27803872, 1000, {0, 0, -10000}, {20000, 0, 0}, {0, 10000, 0}, {0, 0, 10000})
-- MS_setMeshSize(n27803872, GM_entityByTag(r28804384, 2, 2), 1, 0, "500")
MS_setMeshSize(n27803872, GM_entityByTag(r28804384, 2, 2), 1, 0, "500")
-- MS_setGlobalSizeGradationRate(n27803872, 0.25)
MS_setGlobalSizeGradationRate(n27803872, 0.25)
-- AttCase_loadedModel(AttNode_byTag(r29094432, 8))
r28804384 = AttCase_loadedModel(AttNode_byTag(r29094432, 8))
-- AttCase_setModel(AttNode_byTag(r29094432, 1), r28804384)
AttCase_setModel(AttNode_byTag(r29094432, 1), r28804384)
-- AttCase_setModel(AttNode_byTag(r29094432, 2), r28804384)
AttCase_setModel(AttNode_byTag(r29094432, 2), r28804384)
-- PM_new(0, r28804384, 1)
r29032288 = PM_new(0, r28804384, 1)
-- Progress_new()
p29219856 = Progress_new()
-- Progress_setCallback(p29219856, p4597016)
-- Progress_setCallback(p29219856, p4597016)
-- SurfaceMesher_new(n27803872, r29032288)
p29220736 = SurfaceMesher_new(n27803872, r29032288)
-- SurfaceMesher_execute(p29220736, p29219856)
SurfaceMesher_execute(p29220736, p29219856)
-- SurfaceMesher_delete(p29220736)
SurfaceMesher_delete(p29220736)
-- PM_setTotalNumParts(r29032288, 28)
PM_setTotalNumParts(r29032288, 28)
-- VolumeMesher_new(n27803872, r29032288)
p46461232 = VolumeMesher_new(n27803872, r29032288)
-- VolumeMesher_setEnforceSize(p46461232, 0)
VolumeMesher_setEnforceSize(p46461232, 0)
-- VolumeMesher_execute(p46461232, p29219856)
VolumeMesher_execute(p46461232, p29219856)
-- VolumeMesher_delete(p46461232)
VolumeMesher_delete(p46461232)
-- Progress_delete(p29219856)
Progress_delete(p29219856)
-- AttCase_associate(AttNode_byTag(r29094432, 1), nil)
AttCase_associate(AttNode_byTag(r29094432, 1), nil)
-- AttCase_unassociate(AttNode_byTag(r29094432, 1))
AttCase_unassociate(AttNode_byTag(r29094432, 1))
-- MS_deleteMeshCase(n27803872)
MS_deleteMeshCase(n27803872)
-- MS_deleteMeshCase(AttNode_byTag(r29094432, 1))
MS_deleteMeshCase(AttNode_byTag(r29094432, 1))

**Comment by ibaned** [comment]
@rettenbs 

The assert indicates that two MPI processes do not agree on how to partition one of the entities (which MPI ranks have copies of it).

https://github.com/SCOREC/core/blob/master/apf/apfVerify.cc#L256

This is likely an error in the inputs to mesh construction or partitioning.

The most helpful file to see would have been your `src/tools/pumgen.cpp`, but that does not appear to be part of the PUML repository that you indicated.

I recommend searching for the smallest reproducible case, meaning try fewer MPI ranks and a mesh case with fewer elements.

**Comment by rettenbs** [comment]
The code for creating the mesh is here:
https://github.com/TUM-I5/PUML/blob/seissol_converter/src/tools/input/SimModSuite.h#L156

We directly use the Simmetrix mesh and plug it into APF. We only add some tags, but this should not affect the validation.

**Comment by ibaned** [comment]
Thanks Sebastian, I did not see there was a different branch that has the code.
Is Simmetrix generating the mesh in parallel before the first conversion to APF, or does that conversion happen in serial ?
It seems your verification fails right after conversion and alignment, but before re-partitioning with the help of ParMETIS, so we can rule out the latter as an issue.
I'm a bit busy right now but will try later to push a change that prints some information to stderr when this assertion fails.

**Comment by Thomas-Ulrich** [comment]
Hi, thanks for that quick answer.
I hope I wont drown you with my logs, but here is another APF related error (apfMigrate), 

Tue May 10 17:35:19, Info:  Using SimModSuite
Tue May 10 17:35:19, Info:  Loading model
Tue May 10 17:35:30, Info:  Extracting cases
Tue May 10 17:35:30, Info:  Starting the surface mesher
Tue May 10 17:56:28, Info:  Starting the volume mesher
Tue May 10 18:06:30, Info:  Converting mesh to APF
mesh verified in 15.974690 seconds
Tue May 10 18:07:50, Info:  Mesh size: 23381656
Tue May 10 18:07:51, Info:  Minimum insphere found: 2.15896
Tue May 10 18:07:51, Info:  Creating dual graph
Tue May 10 18:07:57, Info:  Creating partitions with METIS
Tue May 10 18:08:09, Info:  Redistributing elements


Abort(1) on node 9 (rank 9 in comm 1140850688): application called MPI_Abort(MPI_COMM_WORLD, 1) - process 9
Thread 11 (Thread 0x2b648c344700 (LWP 38300)):
#0  0x00002b64869f7ce3 in epoll_wait () from /lib64/libc.so.6
#1  0x00002b648bce2436 in poe_exiting_thread () from /opt/ibmhpc/pe1402/base/intel/lib64/libpoe.so
#2  0x00002b6485a17806 in start_thread () from /lib64/libpthread.so.0
#3  0x00002b64869f765d in clone () from /lib64/libc.so.6
#4  0x0000000000000000 in ?? ()
Thread 10 (Thread 0x2b648c545700 (LWP 38302)):
#0  0x00002b64869f7ce3 in epoll_wait () from /lib64/libc.so.6
#1  0x00002b648bce0bb3 in pm_child_sig_thread () from /opt/ibmhpc/pe1402/base/intel/lib64/libpoe.so
#2  0x00002b6485a17806 in start_thread () from /lib64/libpthread.so.0
#3  0x00002b64869f765d in clone () from /lib64/libc.so.6
#4  0x0000000000000000 in ?? ()
Thread 9 (Thread 0x2b648c746700 (LWP 38304)):
#0  0x00002b6485a1f527 in do_sigwait () from /lib64/libpthread.so.0
#1  0x00002b6485a1f5cd in sigwait () from /lib64/libpthread.so.0
#2  0x00002b648bce1ba2 in pm_async_thread () from /opt/ibmhpc/pe1402/base/intel/lib64/libpoe.so
#3  0x00002b6485a17806 in start_thread () from /lib64/libpthread.so.0
#4  0x00002b64869f765d in clone () from /lib64/libc.so.6
#5  0x0000000000000000 in ?? ()
Thread 8 (Thread 0x2b648eb90700 (LWP 38327)):
#0  0x00002b6485a1b66c in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
#1  0x00002b648d5dc58b in hal_ibl_user_intr_hndlr () from /opt/ibmhpc/pe1402/base/intel/lib64/libhal64_ibm.so
#2  0x00002b6485a17806 in start_thread () from /lib64/libpthread.so.0
#3  0x00002b64869f765d in clone () from /lib64/libc.so.6
#4  0x0000000000000000 in ?? ()
Thread 7 (Thread 0x2b648ed91700 (LWP 38328)):
#0  0x00002b64869f7ce3 in epoll_wait () from /lib64/libc.so.6
#1  0x00002b648d5dc183 in hal_ibl_async_intr_hndlr () from /opt/ibmhpc/pe1402/base/intel/lib64/libhal64_ibm.so
#2  0x00002b6485a17806 in start_thread () from /lib64/libpthread.so.0
#3  0x00002b64869f765d in clone () from /lib64/libc.so.6
#4  0x0000000000000000 in ?? ()
Thread 6 (Thread 0x2b649374c700 (LWP 38351)):
#0  0x00002b6485a1b66c in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
#1  0x00002b64896883a3 in shm_dispatcher_thread (arg=0x2b648f19a9ac) at /project/sprelbrew/build/rbrews002a/src/ppe/lapi/lapi_shm.c:2282
#2  0x00002b6485a17806 in start_thread () from /lib64/libpthread.so.0
#3  0x00002b64869f765d in clone () from /lib64/libc.so.6
#4  0x0000000000000000 in ?? ()
Thread 5 (Thread 0x2b6493bf9700 (LWP 38357)):
#0  0x00002b6485a1b66c in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
#1  0x00002b64896a036f in rc_ibl_intr_hndlr (param=0x2b6489cb115c <intr_hndlr_info+60>) at /project/sprelbrew/build/rbrews002a/src/ppe/lapi/lapi_rc_rdma_verbs_wrappers.c:1180
#2  0x00002b6485a17806 in start_thread () from /lib64/libpthread.so.0
#3  0x00002b64869f765d in clone () from /lib64/libc.so.6
#4  0x0000000000000000 in ?? ()
Thread 4 (Thread 0x2b6493dfa700 (LWP 38358)):
#0  0x00002b64869f7ce3 in epoll_wait () from /lib64/libc.so.6
#1  0x00002b64896a07b1 in rc_ibl_async_intr_hndlr (param=0x16) at /project/sprelbrew/build/rbrews002a/src/ppe/lapi/lapi_rc_rdma_verbs_wrappers.c:1374
#2  0x00002b6485a17806 in start_thread () from /lib64/libpthread.so.0
#3  0x00002b64869f765d in clone () from /lib64/libc.so.6
#4  0x0000000000000000 in ?? ()
Thread 3 (Thread 0x2b6493ffb700 (LWP 38360)):
#0  0x00002b6485a1b9fc in pthread_cond_timedwait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
#1  0x00002b648966a37f in _timer_arm (timer=0x17d48a4) at /project/sprelbrew/build/rbrews002a/src/ppe/lapi/intrhndlrs.c:340
#2  0x00002b648966a0e4 in _lapi_tmr_thrd (param=0x17d48a4) at /project/sprelbrew/build/rbrews002a/src/ppe/lapi/intrhndlrs.c:541
#3  0x00002b6485a17806 in start_thread () from /lib64/libpthread.so.0
#4  0x00002b64869f765d in clone () from /lib64/libc.so.6
#5  0x0000000000000000 in ?? ()
Thread 2 (Thread 0x2b64941fc700 (LWP 38385)):
#0  0x00002b64869ee186 in poll () from /lib64/libc.so.6
#1  0x00002b648baac642 in Connection::Wait (this=0x17e2600) at /project/sprelbrew/build/rbrews002a/src/ppe/pnsd/connection.cpp:169
#2  0x00002b648ba983ce in internal_pnsd_api_wait_for_updates (handle=<optimized out>, wakeup_event=0x2b64941ed874, device_name=<optimized out>, adapter_type=<optimized out>, win_id=<optimized out>, cmd_string=0x2b64941ed860, opt_length=0x2b64941ed870, opt=0x2b64941ed868) at /project/sprelbrew/build/rbrews002a/src/ppe/pnsd/pnsd_api.cpp:337
#3  0x00002b648ba986ac in pnsd_api_wait_for_updates (handle=-1809918288, wakeup_event_OUT=0x2, cmd_string_OUT=<optimized out>, opt_length=<optimized out>, opt_OUT=<optimized out>) at /project/sprelbrew/build/rbrews002a/src/ppe/pnsd/pnsd_api.cpp:379
#4  0x00002b648968f0ab in preempt_monitor_thread (param=0x2b64941ed6b0) at /project/sprelbrew/build/rbrews002a/src/ppe/lapi/lapi_preempt.c:523
#5  0x00002b6485a17806 in start_thread () from /lib64/libpthread.so.0
#6  0x00002b64869f765d in clone () from /lib64/libc.so.6
#7  0x0000000000000000 in ?? ()
Thread 1 (Thread 0x2b648ba861a0 (LWP 38289)):
#0  0x00002b64869c3b8f in waitpid () from /lib64/libc.so.6
#1  0x00002b6486957cc1 in do_system () from /lib64/libc.so.6
#2  0x00002b648695806c in system () from /lib64/libc.so.6
#3  0x00002b648bce273f in pm_linux_print_coredump () from /opt/ibmhpc/pe1402/base/intel/lib64/libpoe.so
#4  0x00002b648bce1fda in pm_lwcf_signal_handler () from /opt/ibmhpc/pe1402/base/intel/lib64/libpoe.so
#5  <signal handler called>
#6  0x00002b64869970bc in free () from /lib64/libc.so.6
#7  0x00000000004e379d in mds_set_copies (net=0x119919ac0, m=0x1, e=428972720, c=0x1199a6a90) at /gpfs/work/h019z/di73yeq/core-sim/mds/mds_net.c:62
#8  0x00000000004fe902 in apf::bcastRemotes (m=0x119919ac0, senders=...) at /gpfs/work/h019z/di73yeq/core-sim/apf/apfMigrate.cc:598
#9  0x0000000000501b1a in setupRemotes (m=<optimized out>, received=..., senders=...) at /gpfs/work/h019z/di73yeq/core-sim/apf/apfMigrate.cc:621
#10 moveEntities (m=<optimized out>, senders=<optimized out>) at /gpfs/work/h019z/di73yeq/core-sim/apf/apfMigrate.cc:638
#11 apf::migrate1 (m=0x119919ac0, plan=0x1) at /gpfs/work/h019z/di73yeq/core-sim/apf/apfMigrate.cc:798
#12 0x000000000050306e in migrateSilent (m=<optimized out>, plan=<optimized out>) at /gpfs/work/h019z/di73yeq/core-sim/apf/apfMigrate.cc:841
#13 apf::migrate (m=0x119919ac0, plan=0x1) at /gpfs/work/h019z/di73yeq/core-sim/apf/apfMigrate.cc:846
#14 0x0000000000451c03 in main ()
Abort(1) on node 1 (rank 1 in comm 1140850688): application called MPI_Abort(MPI_COMM_WORLD, 1) - process 1
Abort(1) on node 2 (rank 2 in comm 1140850688): application called MPI_Abort(MPI_COMM_WORLD, 1) - process 2
Abort(1) on node 3 (rank 3 in comm 1140850688): application called MPI_Abort(MPI_COMM_WORLD, 1) - process 3
Abort(1) on node 4 (rank 4 in comm 1140850688): application called MPI_Abort(MPI_COMM_WORLD, 1) - process 4
Abort(1) on node 5 (rank 5 in comm 1140850688): application called MPI_Abort(MPI_COMM_WORLD, 1) - process 5
Abort(1) on node 6 (rank 6 in comm 1140850688): application called MPI_Abort(MPI_COMM_WORLD, 1) - process 6
Abort(1) on node 7 (rank 7 in comm 1140850688): application called MPI_Abort(MPI_COMM_WORLD, 1) - process 7
Abort(1) on node 8 (rank 8 in comm 1140850688): application called MPI_Abort(MPI_COMM_WORLD, 1) - process 8
Abort(1) on node 0 (rank 0 in comm 1140850688): application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0

**Comment by Thomas-Ulrich** [comment]
yes the the mesh was generated in parallel.
(28 CPUs for the first error log, 8 for the other one)

**Comment by ibaned** [comment]
These probably have the same root cause, somehow the APF mesh is being created with bad parallel connectivity information.
As I said before, could you try to create a case that fails with a minimal number of MPI ranks, such that perhaps we can compile your code and attempt to replicate it ?
I suppose 8 ranks is small enough that we could try it, but if you can replicate a failure with 4 or even 2 ranks that would be excellent.
In the meantime, could you point us to compile instructions for this tool (if you think it is reasonable for us to try compiling it on our systems) ? Thank you.

**Comment by Thomas-Ulrich** [comment]
Hi, the first error seems to occur only when using ~20 CPU and more.
The other error log occurs when creating a really large mesh (~30 million elements), but indeed even with less CPUs (so you will need a machine with a lot of memory). 
here is a page describing the compilation steps of PUMGen:
https://github.com/TUM-I5/PUML/wiki/How-to-compile-PUMGen

**Comment by ibaned** [comment]
This may have to do with integer overflow limitations.
Could you try to reconfigure our code with `-DMDS_ID_TYPE=long` ?
Lets see how that does before we try anything else.

**Comment by Thomas-Ulrich** [comment]
hi, yes compiling with -DMDS_ID_TYPE=long option solved most of the problems! Thank you!
The error associated with the first error log I send is still poping out, but as I said, it only occurs  when using ~20 CPU and more. This mesh being rather small (600k elements) it does not makes much sense to use so many CPU anyway. So I close the issue.
Thanks again!

**Comment by rettenbs** [comment]
If you use to many tasks, you might end up with tasks with no elements at all. Maybe this is a problem.

**Comment by ibaned** [comment]
Good, I suspected that would resolve the 30 million elements in serial case. I am surprised that with 600K elements to distribute among only 20 MPI ranks, you get ranks with no elements, but maybe this is just the way Simmetrix outputs the mesh ? As far as I know, this should not break the conversion to APF, but I have never tried it. Feel free to open a separate issue if the verification failures continue to be a problem.
